src/AssemblyGenerator.Types.cs

```diff
@@ -1,95 +1,146 @@
+                if (!_metadata.TryGetTypeDefinition(type, out var metadata))
+                {
+                    throw new InvalidOperationException($"Type definition metadata cannot be found: {type}");
+                }
 
-            return refType;
+                metadata.MarkAsEmitted();
+            }
         }
 
-        internal bool IsReferencedType(Type type)
+        private void ReserveTypes(IEnumerable<Type> types)
         {
-            // todo, also maybe in Module, ModuleRef, AssemblyRef and TypeRef
-            // ECMA-335 page 273-274
-            return type.Assembly != _currentAssembly;
-        }
+            var offset = new TypeDefinitionMetadataOffset
+            {
+                FieldIndex = _metadata.Builder.GetRowCount(TableIndex.Field),
+                PropertyIndex = _metadata.Builder.GetRowCount(TableIndex.PropertyMap),
+                MethodIndex = _metadata.Builder.GetRowCount(TableIndex.MethodDef)
+            };
 
-        internal EntityHandle GetOrCreateType(Type type)
-        {
-            if (_typeHandles.ContainsKey(type.GUID))
+            foreach (var type in types)
             {
-                return _typeHandles[type.GUID];
+                var nextOffset = ReserveTypeDefinition(type, offset);
+                offset = nextOffset;
             }
+        }
 
-            var baseType = default(EntityHandle);
+        private TypeDefinitionMetadataOffset ReserveTypeDefinition(Type type, TypeDefinitionMetadataOffset offset)
+        {
+            var baseTypeHandle = type.BaseType != null ? _metadata.GetTypeHandle(type.BaseType) : default;
+
+            var fieldRowCount = offset.FieldIndex;
+            var propertyRowCount = offset.PropertyIndex;
+            var methodRowCount = offset.MethodIndex;
 
-            if (IsReferencedType(type))
+            foreach (var field in type.GetFields(AllFields))
             {
-                return CreateReferencedType(type);
+                var handle = MetadataTokens.FieldDefinitionHandle(fieldRowCount + 1);
+                _metadata.ReserveFieldDefinition(field, handle);
+                ++fieldRowCount;
             }
 
-            if (type.BaseType != null)
+            foreach (var property in type.GetProperties(AllProperties))
             {
-                var bsType = type.BaseType;
-                if (bsType.Assembly != _currentAssembly)
+                var propertyHandle = MetadataTokens.PropertyDefinitionHandle(propertyRowCount + 1);
+                MethodDefinitionHandle getMethodHandle = default;
+                MethodDefinitionHandle setMethodHandle = default;
+
+                if (property.GetMethod != null)
                 {
-                    var bsTypeRef = CreateReferencedType(bsType);
-                    _typeHandles[bsType.GUID] = bsTypeRef;
-                    baseType = bsTypeRef;
+                    getMethodHandle = MetadataTokens.MethodDefinitionHandle(methodRowCount + 1);
```

Comment:
An alternative coding style that I tend to prefer:

    getMethodHandle = MetadataTokens.MethodDefinitionHandle(++methodRowCount);

Reply:
I especially avoid that style in order to make intention clearer. Because, one can use multiple pre or post increment with in the single assignment which can be confusing. So, I always avoid to pre/post increment in assignments. Regarding why I prefer pre-increment over post-increment for incrementing variables -including for-loops-, it's a habit from old times where compilers produced inefficient code for post-increments 🙂